src/herder/TxSetFrame.h

```diff
@@ -17,105 +16,99 @@ namespace stellar
-
-    virtual int64_t getBaseFee(LedgerHeader const& lh) const = 0;
-
-    virtual Hash const& getContentsHash() = 0;
-
-    virtual Hash const& previousLedgerHash() const = 0;
-
-    virtual size_t sizeTx() const = 0;
-
-    virtual size_t sizeOp() const = 0;
-
-    virtual std::vector<TransactionFrameBasePtr> sortForApply() = 0;
-    virtual void toXDR(TransactionSet& set) = 0;
-};
-
-class TxSetFrame : public AbstractTxSetFrameForApply
-{
-    std::optional<Hash> mHash;
-
-    // mValid caches both the last app LCL that we checked
-    // vaidity for, and the result of that validity check.
-    std::optional<std::pair<Hash, bool>> mValid;
-
-    Hash mPreviousLedgerHash;
-
     using AccountTransactionQueue = std::deque<TransactionFrameBasePtr>;
+    using Transactions = std::vector<TransactionFrameBasePtr>;
 
-    bool checkOrTrim(Application& app,
-                     std::vector<TransactionFrameBasePtr>& trimmed,
-                     bool justCheck, uint64_t lowerBoundCloseTimeOffset,
-                     uint64_t upperBoundCloseTimeOffset);
-
-    UnorderedMap<AccountID, AccountTransactionQueue> buildAccountTxQueues();
-    friend struct SurgeCompare;
-
-  public:
-    std::vector<TransactionFrameBasePtr> mTransactions;
+    // Constructs from a list of transactions. If inputIsApplyOrder is true (in
+    // the case of simulation), the input transactions are already in apply
+    // order, default should be false.
+    TxSetFrame(Hash const& previousLedgerHash, Transactions const& transactions,
+               bool inputIsApplyOrder);
 
     TxSetFrame(Hash const& previousLedgerHash);
 
-    TxSetFrame(TxSetFrame const& other) = default;
-
     // make it from the wire
     TxSetFrame(Hash const& networkID, TransactionSet const& xdrSet);
 
+    TxSetFrame(TxSetFrame const& other) = default;
+
     virtual ~TxSetFrame(){};
 
     // returns the hash of this tx set
-    Hash const& getContentsHash() override;
-
-    Hash& previousLedgerHash();
-    Hash const& previousLedgerHash() const override;
-
-    virtual void sortForHash();
-
-    std::vector<TransactionFrameBasePtr> sortForApply() override;
+    Hash const&
+    getContentsHash() const
+    {
+        return mHash;
+    }
 
-    bool checkValid(Application& app, uint64_t lowerBoundCloseTimeOffset,
-                    uint64_t upperBoundCloseTimeOffset);
+    Hash const&
+    previousLedgerHash() const
+    {
+        return mPreviousLedgerHash;
+    }
 
-    // remove invalid transaction from this set and return those removed
-    // transactions
-    std::vector<TransactionFrameBasePtr>
-    trimInvalid(Application& app, uint64_t lowerBoundCloseTimeOffset,
-                uint64_t upperBoundCloseTimeOffset);
-    void surgePricingFilter(Application& app);
+    Transactions const&
+    getTxsInHashOrder() const
+    {
+        return mTxsInHashOrder;
+    }
 
-    void removeTx(TransactionFrameBasePtr tx);
+#ifdef BUILD_TESTS
+    Transactions&
+    transactions()
```

Comment:
How frequently do we use the mutable method in the tests? I recall only one or two test cases where we directly modify the internal tx vector. Instead, we could just modify the temporary vector and construct new tx sets.

Reply:
I don't see what exactly does this test cover now, given that `TxSetFrame` now maintains the inavriant of tx set being hash-sorted. So the scenario you described just cannot happen anymore.